File: 1487-cinema-seat-allocation/cinema-seat-allocation.py
class Solution:
    def maxNumberOfFamilies(self, n: int, reservedSeats: List[List[int]]) -> int:
        reservedByRow = {}

        for row, seat in reservedSeats:
            if row not in reservedByRow:
                reservedByRow[row] = set()
            reservedByRow[row].add(seat)

        count = 0

        for row, seatsTaken in reservedByRow.items():
            left = 2 not in seatsTaken and 3 not in seatsTaken and 4 not in seatsTaken and 5 not in seatsTaken
            mid = 4 not in seatsTaken and 5 not in seatsTaken and 6 not in seatsTaken and 7 not in seatsTaken
            right = 6 not in seatsTaken and 7 not in seatsTaken and 8 not in seatsTaken and 9 not in seatsTaken

            if left and right:
                count += 2
            elif left or mid or right:
                count += 1

        count += 2 * (n - len(reservedByRow))

        return count

File: 1487-cinema-seat-allocation/cinema-seat-allocation.cpp
class Solution {
public:
    int maxNumberOfFamilies(int n, vector<vector<int>>& reservedSeats) 
    {
        unordered_map<int, unordered_set<int>> reservedByRow;

        for (int i = 0; i < reservedSeats.size(); i++)
        {
            vector<int> s = reservedSeats[i];
            reservedByRow[s[0]].insert(s[1]);
        }

        int count = 0;

        for (auto& [row, seatsTaken] : reservedByRow)
        {
            bool left = !seatsTaken.count(2) && !seatsTaken.count(3) && !seatsTaken.count(4) && !seatsTaken.count(5);
            bool mid = !seatsTaken.count(4) && !seatsTaken.count(5) && !seatsTaken.count(6) && !seatsTaken.count(7);
            bool right = !seatsTaken.count(6) && !seatsTaken.count(7) && !seatsTaken.count(8) && !seatsTaken.count(9);

            if (left && right)
            {
                count += 2;
            }
            else if (left || mid || right)
            {
                count += 1;
            }
        }

        count += 2 * (n - (int)reservedByRow.size());

        return count;
    }
};